index.js

```diff
@@ -18,28 +18,27 @@ module.exports.init = function (options, callback) {
   if (!options.secret) options.secret = process.env.ARTSY_SECRET;
 
   // Fetch the xapp token, cache, and refresh
-  fetchAndCacheToken(options)(callback);
+  fetchAndCacheToken(options, callback);
 }
 
-var fetchAndCacheToken = function (options){
-  return function(callback) {
-
-    // Get the token
-    request
-      .get(options.url + '/api/v1/xapp_token')
-      .query({ client_id: options.id, client_secret: options.secret })
-      .end(function(err, res) {
-        if (err) {
-          module.exports.emit('error', err);
-          if (callback) callback(err);
-          return
-        }
-        module.exports.token = res.body.xapp_token;
-        if (callback) callback(null, module.exports.token);
-
-        // Recurse this function to refresh the token it before it expires
-        var expiresAt = new Date(res.body.expires_in).getTime();
-        setTimeout(fetchAndCacheToken(options), (expiresAt - 1000) - Date.now());
-      });
-    }
+var fetchAndCacheToken = function (options, callback) {
+  // Get the token
+  request
+    .get(options.url + '/api/v1/xapp_token')
+    .query({ client_id: options.id, client_secret: options.secret })
+    .end(function(err, res) {
+      if (err) {
+        module.exports.emit('error', err);
+        if (callback) callback(err);
+        return
+      }
+
+      // Set and callback w/ token (useful for client middleware).
+      module.exports.token = res.body.xapp_token;
+      if (callback) callback(null, module.exports.token);
+
+      // Recurse this function to refresh the token it before it expires
+      var expiresAt = new Date(res.body.expires_in).getTime();
+      setTimeout(() => fetchAndCacheToken(options, callback), (expiresAt - 1000) - Date.now());
```

Comment:
In my opinion this should be more conservative than scheduling the callback for just 1 second before its expiration. That kind of scheduling isn't bulletproof, and if this fails or gets delayed slightly the effect is catastrophic. I'd opt for refreshing the token at 75-80% of expiration instead.

Reply:
Yea, that's true. I like that 75% idea.